### PR TITLE
feat: jsonld parsing without re-serialization

### PR DESCRIPTION
**Pull request by jeswr**
Note that the custom caching mechanism here is only temporary. There is work upstream to:
 - Improve the speed of parsing JSON-LD contexts by 10x or more (https://github.com/rubensworks/jsonld-context-parser.js/pull/73).
  - Add the caching mechanism upstream https://github.com/rubensworks/jsonld-context-parser.js/pull/70

### CODE REVIEW

**Review by NSeydoux — APPROVED** [review]

File: src/parser/jsonld.ts
Comment (jeswr) [reply src/parser/jsonld.ts:73]: `Object.keys` always returns a `string[]` even if the set of keys defined in the type is more restricted. So in 73 there is a type error because `key` is a `string` but `opts` only accepts a finite set of keys.

File: src/lookup/query.ts
Comment (jeswr) [reply src/lookup/query.ts:158]: This is to avoid memory overflow errors from trying to parse too many VCs in parallel.

Granted this shouldn't be an issue with all of the caching and optimisation work that has taken place; I think we should still keep it as there is still a risk of memory errors if, say, millions of VCs are being processed.

I should also note that this doesn't really slow down processing given that we are in a single-threaded environment and most of the operations taking place in the parsing step are synchronous.

File: src/lookup/query.ts
Comment (jeswr) [inline src/lookup/query.ts:152]:
```suggestion
      newVerifiableCredential.push(
        // Limit concurrency to avoid memory overflows. For details see
        // https://github.com/inrupt/solid-client-vc-js/pull/849#discussion_r1377400688
        // eslint-disable-next-line no-await-in-loop
        ...(await Promise.all(
```

File: src/lookup/query.ts
Comment (jeswr) [inline src/lookup/query.ts:148]:
```suggestion
    const newVerifiableCredential: NonNullable<ParsedVerifiablePresentation['verifiableCredential']> = [];
```

File: src/lookup/query.ts
Comment (jeswr) [reply src/lookup/query.ts]: Done

File: src/lookup/query.ts
Comment (jeswr) [reply src/lookup/query.ts]: I've added a comment to the code to reflect this.

File: src/lookup/query.ts
Comment (jeswr) [inline src/lookup/query.ts:148]:
```suggestion
    const newVerifiableCredential: NonNullable<ParsedVerifiablePresentation["verifiableCredential"]> = [];
```

File: src/parser/contexts/status-list.ts
Comment (NSeydoux) [inline src/parser/contexts/status-list.ts:21]: Can you add the context URL as a comment to all of them? It's pretty convenient.

**Comment by NSeydoux** [comment]
These changes are being tested in the access-grant library, until they are validated we'll hold off merging this PR.

File: src/common/getters.ts
Comment (NSeydoux) [inline src/common/getters.ts:29]: The pattern we've used in `solid-client` is to define a `WithId` type, and not necessarily have a `DatasetWithId` but rather a `Datasetcore & WithId`. Arguably, that could be aliased to keep `DatasetWithId`.

This allows to be a bit more fine-grained: for instance, `getId` only needs a `withId` typed argument.

File: src/lookup/query.ts
Comment (NSeydoux) [inline src/lookup/query.ts:82]: The orders of declaration in the file will be used for the order of appearance in the API docs, so you'll want to move the deprecated functions after the non-deprecated ones.

File: src/common/common.ts
Comment (jeswr) [reply src/common/common.ts:581]: Note that this change will result in errors like `cannot read property proof of undefined` rather than `Parsing the Verifiable Credential [http://some.vc/] as JSON failed: ${cannot read property proof of undefined}`.

If you are ok with this dropping of context in the error message then we can go ahead with this change.

File: src/issue/issue.ts
Comment (jeswr) [reply src/issue/issue.ts:63]: If we don't have a type overload for `boolean` and only one type overload for `true` and one type overload for `false` then we end up with a type error when the value passed in is of type `boolean` (in this case the input of `issueVerifiableCredential`  which calls `internal_issueVerifiableCredential` will be `boolean` in the input signature rather than explicitly true or false.).

Instead we just put the `?: true` case first and mark that as deprecated and then everything else is `?: boolean` because the type overloads are evaluated in order of appearance so.

Note this ordered evaluated of type also makes https://github.com/inrupt/solid-client-vc-js/pull/849#discussion_r1405936803 a little harder because of the fact that we want to catch the `?: true` case and then catch everything else with `?: boolean`.

My proposal to resolve this and https://github.com/inrupt/solid-client-vc-js/pull/849#discussion_r1405936803 is that we have 3 overloads in the following order.

`: false`: Not deprecated and used for the JSDoc
`?: true`: Marked as deprecated
`?: boolean`: Used to catch when cases there is a boolean parameter that is not yet known to be `true` or `false`. Possibly mark as deprecated as it does have the possibility of being `true`.

File: src/common/common.ts
Comment (NSeydoux) [reply src/common/common.ts:581]: You're right, the former is clearer. Let's keep things as they are, but maybe add a comment to explain, because I'll forget about this and want to have the same change in 6 months time ^^

File: src/issue/issue.ts
Comment (NSeydoux) [reply src/issue/issue.ts:63]: Sounds good.

**Comment by NSeydoux** [comment]
Beware, this introduced circular dependencies. Rollup is complaining about it, let's make sure to fix it before release.

File: src/common/common.test.ts
Comment (NSeydoux) [inline src/common/common.test.ts:668]: Why is this disabled?

Comment (jeswr) [reply src/common/common.test.ts:668]: Because this would be a new check (the code for this is also currently deleted and/or disabled) - I suggest we enabled it in the next major version.

File: src/common/common.test.ts
Comment (jeswr) [reply src/common/common.test.ts:353]: When this is parsed as JSON-LD it is treated as there being no verifiable credentials; so this is a valid presentation in RDFJS; but when treated as json we have an empty object.

So in this case I would actually expect them to return different results.

If that makes sense, thumbs up this comment and I'll comment this on the test.

File: src/common/common.test.ts
Comment (jeswr) [reply src/common/common.test.ts:399]: This case is similar to https://github.com/inrupt/solid-client-vc-js/pull/849#discussion_r1414119041. The parser will ignore the holder because the context requires it to be an IRI.

So in RDFJS there is no holder triple; and since holder is allowed to be undefined this function returns true.

Similar to https://github.com/inrupt/solid-client-vc-js/pull/849#discussion_r1414119041, if this makes sense to you, thumbs up this comment and I will add a comment.

File: src/index.ts
Comment (jeswr) [reply src/index.ts:36]: Correct - I expect the only consumers of this function to be us in the access-grants library, and I want us to have the liberty to remove it at any time.

**Pull request merged**
